### ISSUE TITLE
GMeans run into dead loop

### ISSUE THREAD

**Issue by luciferlu** [haifengl/smile#183]
Hi,
I am using GMeans cluster in my application, but I found a data sample which will run into the dead loop. I am not sure if this is expected behaviour or something wrong with GMeans implementation.
Below is the code I run:
```
    val data = Array(
      Array(0.0d, 0.344039311677285d),
      Array(0.0d, 1.109973926053626d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, 0.24829748488024242d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, 1.109973926053626d),
      Array(0.0d, 1.109973926053626d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, 1.4929412332417966d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, 0.344039311677285d),
      Array(0.0d, 0.344039311677285d),
      Array(0.0d, 0.344039311677285d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, 0.008942917887635819d),
      Array(0.0d, 0.15255565808319976d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, 1.4929412332417966d),
      Array(0.0d, -0.23041164910497078d),
      Array(0.0d, 1.109973926053626d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.0389279955108855d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, 1.109973926053626d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.23041164910497078d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.23041164910497078d),
      Array(0.0d, 1.8759085404299671d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3740243893005347d),
      Array(0.0d, 0.344039311677285d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.23041164910497078d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.0389279955108855d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, 1.109973926053626d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, 8.194869109034782d),
      Array(0.0d, 0.7270066188654556d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d),
      Array(0.0d, -0.3261534759020134d)
    )

    val gMeans = new GMeans(data, 8)

    println(gMeans)
```

**Comment by haifengl** [comment]
It is a bad data. All the first column is 0. The data is essentially one dimensional. Kmeans and alike won't work on one dimensional data.

**Comment by luciferlu** [comment]
Do you mean input data should preprocess, and exclude all this kind of data set? I am not sure how to distinguish the data not fit GMeans. For example:
```
    val data = Array(0.1d, 0.0d) :: (0 to 30)
      .map(_ => Array(0.0d, 0.1d))
      .toList
    val gMeans = new GMeans(data.toArray, 100)

```
I know this is also not a good data, But it is harder to distinguish than the first example. This code will also run into the dead loop.

**Comment by haifengl** [comment]
Why don't you use some real data? Even if you have to use synthetic data, you have to at least follow some distributions. Kmeans and alike work well on Gaussian distribution.

Note that no clustering algorithms work well on all kinds of data. I don't feel your generated data have practical meaning.

**Comment by luciferlu** [comment]
Actually, both data examples are real data. We partition our data into 30k groups, and I found most groups run well, but some groups run into the deal loop. So I tracked into each group and found first data set. I tried to limit iterations in GMeans and passed the first data set, but got an ArrayIndexOutOfBoundsException exception in another data group, which only have 50 pairs of zero. I understand these data are not good enough for this algorithm, but I have to run it against all data groups and I cannot ensure all data have good quality. I believe this is the hardest part of data processing.
